### PR TITLE
contracts: Fix `seal_call` weights

### LINKED ISSUE #9066: Add proper weight to new version of `seal_call`
`"seal1" seal_call` uses the weights of the old API version. This is OK because it is basically the same code. However, extra weight must be charged when the inputs are cloned rather than forwarded.

### PR DESCRIPTION
**Pull request by athei**
Fixes #9066

* Fix a benchmark bug where the weight for copying back the output buffer on `call` and `instantiate` was counted double.
* Charge a surcharge when using `SEAL_CLONE_INPUT` vs `SEAL_FORWARD_INPUT` based on the cloned input.
* Charge a surcharge for `instantiate` only when balanced is transferred (it is now optional to transfer balance).
* Reduce the number of benchmark components for trivial cases (where we just copy).

Drive by fixes (in separate commits):
* Remove stale and superfluous comments describing return types
* `decrement_refcount` should be infallible
* Refcount functions shouldn't be directly exposed to exec.rs (introduces by `set_call` PR)

### CODE REVIEW

File: frame/contracts/src/wasm/runtime.rs [hunk 1]
```diff
@@ -803,8 +811,11 @@ where
 		salt_len: u32,
 	) -> Result<ReturnCode, TrapReason> {
 		self.charge_gas(RuntimeCosts::InstantiateBase { input_data_len, salt_len })?;
-		let code_hash: CodeHash<<E as Ext>::T> = self.read_sandbox_memory_as(code_hash_ptr)?;
 		let value: BalanceOf<<E as Ext>::T> = self.read_sandbox_memory_as(value_ptr)?;
+		if value > 0u32.into() {
+			self.charge_gas(RuntimeCosts::InstantiateSurchargeTransfer)?;
+		}
```
Comment (HCastano) [inline frame/contracts/src/wasm/runtime.rs:817]: My bank does the same thing 😆

**Review by HCastano — COMMENTED** [review]
I need to take another look at the benchmark changes, rest look good though

**Review by cmichi — APPROVED** [review]

File: frame/contracts/src/wasm/runtime.rs [hunk 2]
```diff
@@ -136,6 +136,10 @@ pub enum RuntimeCosts {
 	/// Charge the gas meter with the cost of a metering block. The charged costs are
 	/// the supplied cost of the block plus the overhead of the metering itself.
 	MeteringBlock(u32),
+	/// Weight charged for copying data from the sandbox.
+	CopyIn(u32),
+	/// Weight charged for copying data to the sandbox.
+	CopyOut(u32),
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/runtime.rs:142]: It depends what your point of view is. Those are written from the point of view of the pallet-contracts and not the contract. But I should rename them to make them independent of that point of view.

**Review by HCastano — APPROVED** [review]
Thanks for the clarifications!

**Pull request merged**